Refactor code for building the documentation

fingolfin opened the pull request:
Instead of patching Documenter.jl, simply copy all relevant .md files (and any
other documentation sources, such as images) from AbstractAlgebra/Nemo/Hecke
into the `docs/src` directory of Oscar.jl

Seems to work well enough for me, locally. Let's see what CI makes of it.

Closes #523
Closes #525

fingolfin committed:
Refactor code for building the documentation

Instead of patching Documenter.jl, simply copy all relevant .md files (and any
other documentation sources, such as images) from AbstractAlgebra/Nemo/Hecke
into the `docs/src` directory of Oscar.jl

thofma commented:
I share Claus question about the immutability of julia package directories. I guess it is fine, because building the directory must mutate the package directory anyway (and therefore should happen only for dev'ed packages in principle).

fingolfin commented:
The advantage of this approach is that it is *much* simpler to understand what's going on, and we are not mucking with Documenter.jl internals that might change at any point without a warning.

That said, I think in the long run, the Julia devs would like to make packages director read-only -- but at this point, this completely hypothetical and far too many packages (including some of our own) still rely on writing there, so we'll have plenty of time to react to that, should it ever come.

Note that in particular `Documenter.jl` itself right now writes its output into the package directory...

Anyway, the fix then will be easy enough: we simply adjust the copying code to *also* copy our `docs/src` dir, and copy it all into a temp directory, then use `makedocs (source="path/to/tmpdir", ...)`

fingolfin commented on docs/make_work.jl:
```diff
@@ -1,147 +1,22 @@
-    # Markdown files, however, get added to the document and also stored into
-    # `mdpages`, to be used later.
-    mdpages = String[]
+# Copy documentation from Hecke, Nemo, AnstratAlgebra
```
Perhaps this should also be in `doit()`? In fact, almost everything on the top level could be moved into `doit`, so that we benefit from `Revise`.

fingolfin commented on docs/make_work.jl:
```diff
@@ -160,38 +35,17 @@ end
+const nemo = "Nemo"
+const aa = "AbstractAlgebra"
 
 bib = CitationBibliography(joinpath(@__DIR__, "oscar_references.bib"), sorting = :nyt)
```
I guess `bib` should also be set in `doit`, so that changes to the bibliography get reflected?

thofma commented:
I am happy with this approach and have no opinion on the two questions above.

fingolfin committed:
Get rid of docs/make_local.jl

thofma commented:
I think this should remove the things it copied after the documentation was built. Just to make sure that no one tries to edit the things in `docs/src`.

fingolfin committed:
docs: reload the bibliography each time

fingolfin committed:
docs: hard code dir names

fingolfin committed:
docs: add comments, remove copied files, copy them in doit()

fingolfin commented:
I've revised this and added a bunch of improvements. `docs/make_local.jl` is gone now; the bibliography is re-read in `doit`, and the files are copied by `doit`, too; and now they are also removed before/after we generate the docs, to ensure we don't get tainted copies, and to hopefully prevent people from editing the copies by mistake.
